src/Utopia/Messaging/Adapter.php

```diff
@@ -71,31 +71,134 @@ protected function request(
             \curl_setopt($ch, CURLOPT_POSTFIELDS, $body);
         }
 
-        \curl_setopt($ch, CURLOPT_URL, $url);
-        \curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
-        \curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
-        \curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
-        \curl_setopt($ch, CURLOPT_USERAGENT, "Appwrite {$this->getName()} Message Sender");
+        \curl_setopt_array($ch, [
+            CURLOPT_CUSTOMREQUEST => $method,
+            CURLOPT_URL => $url,
+            CURLOPT_HTTPHEADER => $headers,
+            CURLOPT_RETURNTRANSFER => true,
+            CURLOPT_USERAGENT => "Appwrite {$this->getName()} Message Sender",
+            CURLOPT_TIMEOUT => $timeout,
+        ]);
 
         $response = \curl_exec($ch);
+
         \curl_close($ch);
 
-        if (\curl_errno($ch)) {
-            throw new \Exception('Error: '.\curl_error($ch));
+        try {
+            $response = \json_decode($response, true, flags: JSON_THROW_ON_ERROR);
+        } catch (\JsonException) {
+            // Ignore
         }
 
-        $jsonResponse = \json_decode($response, true);
+        return [
+            'url' => $url,
+            'statusCode' => \curl_getinfo($ch, CURLINFO_RESPONSE_CODE),
+            'response' => $response,
+            'error' => \curl_error($ch),
+        ];
+    }
 
-        if (\json_last_error() == JSON_ERROR_NONE) {
-            return [
-                'response' => $jsonResponse,
-                'statusCode' => \curl_getinfo($ch, CURLINFO_HTTP_CODE),
+    /**
+     * @param  array<string>  $urls
+     * @param  array<string>  $headers
+     * @param  array<string>  $bodies
+     * @return array<array{url: string, statusCode: int, response: array<string, mixed>|null, error: string|null}>
+     *
+     * @throws \Exception
+     */
+    protected function requestMulti(
```

Comment:
some more documentation around the different methods in the adapter would be helpful.

Reply:
Updated docs